src/database/postgres/mapsavequeue.cpp

```diff
@@ -0,0 +1,153 @@
+
+#include "config.h"
+
+#ifdef USE_POSTGRESQL
+
+#include "mapsavequeue.h"
+
+#include "log.h"
+#include "threading/mutex_auto_lock.h"
+#include "exceptions.h"
+
+#include <thread>
+#include <chrono>
+#include <list>
+#include "pgutil.h"
+
+MapSaveQueue::MapSaveQueue(PGconn *m_conn) :
+	Thread("map_save_queue"),
+	m_conn(m_conn) {
+
+	m_pgversion = PQserverVersion(m_conn);
+
+	if (m_pgversion < 90500) {
+		PGUtil::prepareStatement(m_conn, "write_block_insert",
+			"INSERT INTO blocks (posX, posY, posZ, data) SELECT "
+				"$1::int4, $2::int4, $3::int4, $4::bytea "
+				"WHERE NOT EXISTS (SELECT true FROM blocks "
+				"WHERE posX = $1::int4 AND posY = $2::int4 AND "
+				"posZ = $3::int4)");
+
+		PGUtil::prepareStatement(m_conn, "write_block_update",
+			"UPDATE blocks SET data = $4::bytea "
+				"WHERE posX = $1::int4 AND posY = $2::int4 AND "
+				"posZ = $3::int4");
+	} else {
+		PGUtil::prepareStatement(m_conn, "write_block",
+			"INSERT INTO blocks (posX, posY, posZ, data) VALUES "
+				"($1::int4, $2::int4, $3::int4, $4::bytea) "
+				"ON CONFLICT ON CONSTRAINT blocks_pkey DO "
+				"UPDATE SET data = $4::bytea");
+	}
+
+}
+
+MapSaveQueue::~MapSaveQueue(){}
+
+void *MapSaveQueue::run(){
+	std::vector<QueuedItem*> save_items;
+
+	while (!stopRequested()){
+		std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
i don't agree with that usage. Please use semaphores to wake up the thread based on events instead